libs/gltfio/src/ResourceLoader.cpp

```diff
@@ -231,7 +233,13 @@ static void decodeDracoMeshes(FFilamentAsset* asset) {
 }
 
 ResourceLoader::ResourceLoader(const ResourceConfiguration& config) :
-        mPool(new AssetPool), pImpl(new Impl(config)) { }
+        mPool(new AssetPool), pImpl(new Impl(config)) {
+    pImpl->mJobSystem = JobSystem::getJobSystem();
+    if (!pImpl->mJobSystem) {
+        pImpl->mJobSystem = new JobSystem();
```

Comment:
so where is this JobSystem destroyed?

Reply:
I thought about that and decided not to bother, since this way it gets re-used over subsequent runs.

If I destroy it, then I think I'll need to add an ownership bool to resource loader since it may or may not own it.  What do you think?

Reply:
(It's also a little tricky to destroy it because decoding jobs can last longer than the ResourceLoader itself.)

If we decide that it's ok to not bother destroying it, I will at least add a comment explaining why.

Reply:
But what if the app keeps creating and destroying ResourceLoader (for some reason -- maybe because of unit tests), wouldn't it leak a JobSystem each time?

Reply:
Also we don't have an Engine we could get the `JobSystem` from? We  could add a getJobSystem() on Engine?

Reply:
Getting the JobSystem from the `Engine` seems like the safest solution. I'm not opposed to gltfio having its own but the lifecycle needs to be fixed.

Reply:
Re: creating and re-destroying, note that it re-uses the TLS singleton if it is available.

I really like the idea of exposing the Engine's `JobSystem`.  Let me give that a try.

Reply:
It looks like I need to increase the "adoptable" thread count in Engine's JobSystem from 1 to 2.  Mathias does that sound ok to you?  (If I don't call adopt, I hit the assert in getState.). To be honest I'm a little fuzzy on what the adopt / emancipate stuff means, can you refresh my memory?

Reply:
JobSystem adopts the main thread to make it participate in the pool. Jobs can only be created from a thread owned by JS. It should work if you use JS from the "main" thread in gltfio.

Reply:
I'm calling from the main thread (I verified this by dumping the tid) but the assert in `getState` triggers anyway because it relies on a global TLS variable.

Reply:
Is it the adoption hasn't happened yet then?

Reply:
Engine adopts the JobSystem early on, then gltfio tries to use it from the main thread at a later point in time, but this fails because JobSystem's `sThreadState` is null in gltfio's address space.

I can "fix" this by making ResourceLoader call adopt(), but that feels wrong.

Reply:
Gah :( Static objects really are a pain with multiple libraries. How is this normally handled?

Reply:
ah... crap.  I'm sure that adopting twice the same thread will cause issues. I think what will actually happen is that you'll get 2 different ThreadState for the same thread and some jobs won't be executed.
It's a major PITA.  Basically, even though you run in the same thread, there is no way for us to get to the correct TLS slot.

I think to fix this we'll have to use a hashmap (threadid->state) instead of TLS. Do you want to give it a try?  I would do that in a separate change, replace the TLS stuff with a hashmap in JobSystem. I think once you have that, it'll be able to see that the thread is adopted.

Reply:
I think I get it -- to summarize, I will replace `sThreadState` with `mThreadStates`, which will be a `robin_map` from `std::thread::id` to `ThreadState`.

I think this implies that we'll also remove the static getter method `JobSystem::getJobSystem()`.  Looks like there aren't many users of that method, so it should be tractable.

Reply:
Mathias would the hash map that you propose still live in static memory?  Also, just for clarity, what was the purpose of `sThreadState`?

Reply:
Yes that's correct. Currently `getState()` is static because it didn't need to not be -- just make it non static and use the hashmap.
sThreadState is per-thread pointer to a ThreadState, we won't need it anymore because we have the hashmap.

The hashmap would NOT be in static memory, it would be a regular member of JobSystem. In fact there should not be *any* static members at all -- that's what causes us harm.